### DISCUSSION TITLE
Can I create a button to display the map ?

### DISCUSSION THREAD

**Discussion by BuzzLEquerre** [flareteam/flare-game#965] [General]
Hello, I would like to be able to display the map the player is in via a button, is this possible?

Or make sure to display the mini map with a button and name a permanent display.

I saw this on Exiled Kingdom.
If you have an idea of how to add a custom button for the map that would help me a lot 👍

**Comment by dorkster** [comment]
The only map provided by the game is the minimap. If you want a "world" or "regional" map, you can accomplish this by using a MenuBook with the map image as the background. You can then add buttons for places the player can teleport to (if that applies to your game). Ghostlore did this, so here's its `books/map3.txt` file:

```
pos=0,0,640,360
align=center
close=604,12
background=images/menus/backgrounds/map3.png
soundfx_open=soundfx/world_map.ogg
soundfx_close=soundfx/inventory/inventory_close.ogg

[event_open]
music=music/world_map_theme.ogg

[event_close]
music=music/town_theme.ogg

[button]
button_pos=306,68
button_image=images/menus/map_marker.png
cutscene=cutscenes/chapter_01_tip.txt
intermap=maps/village_burning.txt,140,99

[button]
button_pos=338,169
button_image=images/menus/map_marker.png
cutscene=cutscenes/chapter_02_tip.txt
intermap=maps/cemetery.txt,29,170

[button]
button_pos=407,107
button_image=images/menus/map_marker.png
cutscene=cutscenes/chapter_03.txt
intermap=maps/plantation.txt,122,138
```

----

You wouldn't be able to map this to a button directly, but you could do the following:

1. Make a script that opens the book file
2. Create a Power and use `script` to call the script file
3. Create an Item with the `power` attribute pointing to your Power ID.
4. For each of your starting classes, give the player the Item and put the Power on the actionbar

Then, players can use the actionbar hotkey to open the map

**Reply by BuzzLEquerre** [reply]
I created a "world_map" script to open the book file "world_map.txt"
and I added an item that uses this script but once in the game the item doesn't seem to work :

![Capture d'écran 2024-08-16 211654](https://github.com/user-attachments/assets/17fa9bc0-e981-4ebf-89a4-10acb9cbb234)
![Capture d'écran 2024-08-16 212235](https://github.com/user-attachments/assets/12ae5854-d0e3-4c0d-a055-ab6625504969)

I think the problem comes from my script file, only I don't know any other function than INCLUDE

**Reply by dorkster** [reply]
You're close!

First, the script file should contain the following:
```
[event]
book=books/world_map.txt
```

For more information about events, see the EventManager section of the attribute reference: https://github.com/flareteam/flare-engine/wiki/Attribute-Reference#eventmanager

In powers.txt, make sure that the script is pointing to your "map_button.txt" script (it's pointing to dev_tome.txt right now).

In classes.txt, you need to fix 2 things:
- Add `2003` to the `carried` list
- Change the power ID in `actionbar` to `504`, which is the power ID that triggers your script

**Reply by BuzzLEquerre** [reply]
The book appears, but it is impossible to open it, I made the necessary changes, so the problem is in the file
![Capture d'écran 2024-08-16 214526](https://github.com/user-attachments/assets/19c3874a-8c27-4c8b-bca2-a7047813d35b)
![Capture d'écran 2024-08-16 214915](https://github.com/user-attachments/assets/d0d44d88-88fa-4ee1-a155-04ef1502fc07)

I also replaced the music probes and images of the "worldmap.txt" file with my images

**Reply by dorkster** [reply]
Oh I missed something. In `powers/powers.txt`, you need add the following to the power so that it can actually cast:

```
type=fixed
new_state=instant
```

Also, if you haven't changed anything in `books/world_map.txt`, you'll probably only see a couple of blank buttons because those images are from Ghostlore, not flare-game.

**Reply by BuzzLEquerre** [reply]
it works !!
![Capture d'écran 2024-08-17 091502](https://github.com/user-attachments/assets/320e3097-6ee2-4b9d-955c-a2715816f671)
 All you have to do is edit everything xddd thk you !
And it would be possible to make the buttons invisible ?

**Reply by BuzzLEquerre** [reply]
I have a problem on my card, it has the original dimension (3840; 2160) and once in the game it does not appear in full :
- I'd like to know if I can move the map in game with the cursor, and therefore how to do it

- Or if it's possible to place my entire map on the full screen to directly see all the areas

the problem :
![Capture d'écran 2024-08-17 121738](https://github.com/user-attachments/assets/13106ea2-ebf2-430b-994b-0170a6657346)

**Reply by dorkster** [reply]
- You can make the buttons invisible by making an empty, transparent PNG. The dimensions will determine the clickable area of the button. The height should be 4 times the width, since buttons have 4 states: normal, hover, pressed, disabled.
- There's no way to move the map around in this context, so you'd need to scale the image down to fit the resolution that the game uses. If you have a large world map, you may want to divide it up into multiple maps. You can then add buttons to each map to switch between the regions.

**Reply by BuzzLEquerre** [reply]
I  cut my map into 4 parts, and if I make buttons that change the background to display the rest of the map, can't it change the buttons for the new areas ?
![Capture d'écran 2024-08-17 220819](https://github.com/user-attachments/assets/edf14f10-68cf-4cdb-8183-bc947b9c5c95)

And how do I set my new button to equip it ?

**Reply by dorkster** [reply]
You should have 4 separate books, one for each map section. That way, you can define the button positions for match the background image for each map. In addition to the buttons for the points of interest, each map could have 2 buttons to switch the map to the adjacent section.

```
# North-west map: books/map_NW.txt

# go east
[button]
button_pos=...
button_image=images/menus/buttons/right.png
book=books/map_NE.txt

# go south
[button]
button_pos=...
button_image=images/menus/buttons/down.png
book=books/map_SW.txt
```

In order to make the map open to the correct region when you use the item, you'll need to use campaign statuses to track which region the player is in. For each map (meaning the map you make in Tiled), you'd need an `on_load` event to `set_status` for the region and `unset_status` for the other regions. Then in your `map_button.txt` script, you can check the status to load the appropriate map:

```
[event]
requires_status=map_NW
book=books/map_NW.txt

[event]
requires_status=map_NE
book=books/map_NE.txt

etc...
```

**Reply by BuzzLEquerre** [reply]
I did it !
![Capture d'écran 2024-08-18 085105](https://github.com/user-attachments/assets/836ff2ca-80dc-4c5f-982a-5e082f93de1c)

and I would like to display the area when the player clicks on it, is it possible to use a cutscene to display the name and an image of the area in question when the player clicks on it?

**Reply by BuzzLEquerre** [reply]
Good evening, I worked on the world map and I managed to make the system of describing the zones thanks to your advice, only I would like to know if it is possible to lock the button when I select an area so that the player sees the area he has selected. Is this po
![Capture d'écran 2024-12-09 194234](https://github.com/user-attachments/assets/10af80ee-5674-4477-9e3e-4337c0cf5e2a)
ssible ? 
[NE.txt](https://github.com/user-attachments/files/18066414/NE.txt)
![map](https://github.com/user-attachments/assets/bf7c4179-4223-43fa-98b6-126bfe232815)

**Reply by dorkster** [reply]
Buttons can be enabled/disabled with `requires_status` and `requires_not_status`, just like text and image blocks. The main difference is that the button will still be visible if it's disabled (it will use the graphics for the disabled state).

Just to clarify: Which button do you want to "lock"? The square map tile that was clicked on?

**Reply by BuzzLEquerre** [reply]
yes I would like to lock the tile buttons, so that the player sees which tile it is.
